*: rewrite update-ssh-keys in rust

sdemos opened the pull request:
This replaces both the authorized_keys.d library and the command line tool.

sdemos committed:
*: cargo project scaffolding

this includes a brand new Cargo.toml and a bunch of boilerplate files.

sdemos committed:
*: implemented update-ssh-keys

this is a wholesale reimplementation of the authorized_keys_d library
and the update-ssh-keys command-line utility, entirely in rust

sdemos committed:
main: removed debugging print statements

we want to try and be closer to the bash script implementation in terms
of output, and the debug statements were showing a lot of implementation
details.

sdemos committed:
cargo: optimize release binary size

the command line utility is included in every install of container
linux, so we want it to be relatively small.

sdemos committed:
cargo: use published openssh-keys package

this makes it easier to refer to the dependency, instead of pointing
directly to the github version.

sdemos committed:
cargo: remove panic=abort from release build

it's easier to build it in the sdk with the way we are building the rust
standard libraries if everything does panic=unwind.

lucab committed:
lib: clippy cleanup

we ran clipply on the source code. this is the result.

lucab committed:
bin: clippy cleanup

we ran clippy on the command line utility and applied the
recommendations.

lucab committed:
main: remove one spurious todo

this shows the user that the command line utility is going to operate on
in the help message

lucab committed:
cargo: bump error-chain to 0.11

This updates error-chain to 0.11, cleaning a spurious clippy warning.
This also opts out from the default backtrace-sys feature, which can
be enabled by the consuming application instead.

sdemos committed:
lib: fixed logging in write function

when we remove/create the staging directory, it printed out the actual
folder instead, even though it was acting on the staging one.

sdemos committed:
lib: fix user switching function

for now, we are implementing the guard manually, because of
ogham/rust-users#16. If that gets merged, we can go back to just using
switch_user_group.

sdemos committed:
lib: move read_keys function to openssh_keys

the read_keys function is generally useful for people. in particular,
coreos-metadata used it from here. it makes more sense to be a part of
that library.

sdemos committed:
users: update rust-users dependency

ogham/rust-users#16 got merged and then they released a patched version
to crates.io. switch back to that.

sdemos committed:
lib: harden directory manipulation code

broke directory replacing and creation into separate functions, which
now explicitly deal with non-existant paths and other filesystem
gotchas.

sdemos committed:
*: delete go

we are rewriting the project in rust, so the first step is to remove all
the old go code

sdemos commented:
/cc @crawford

crawford commented:
Travis config?

sdemos committed:
travis: added travis configuration for rust

this tests it on all the rust compilers, although we can fail on the
nightly compiler.

sdemos committed:
tailor: added tailor config

this lets us automate the checking of commit messages to make sure they
actually conform to our contributing guide.

sdemos committed:
contributing: add rust style recommendation

instead of listing the go style guide, which this replaces, we say that
you should be using rustfmt to format your code.

lucab commented:
It looks like ignition has a reference to this package in its glide.yaml. Should we maybe first move the existing go package over there?

sdemos commented:
I was thinking that we were just going to continue to refer to the old commit in the tree, but talking to derek, he pointed out that then we wouldn't be able to patch it if we find a bug. Instead, since ignition is the only other consumer of the library portion of this project, I opened a pr to move it under there (coreos/ignition#467)

euank commented on .gitignore:
```diff
@@ -1,3 +1,3 @@
-*.swp
+/target/
+**/*.rs.bk
+Cargo.lock
```
Why aren't we checking in `Cargo.lock`?

sdemos replied:
we can. iirc this is just how cargo set it up when I started the project. maybe that's one of the `cargo new --bin` vs `cargo new --lib` differences.

euank replied:
It is; `bin` leaves it unignored.

Isn't this a `bin` not a `lib` really though?

sdemos replied:
it's a little of both. it's the command line utility but also a library consumed by `coreos-metadata`. I can include it.

bgilbert commented:
Have you checked that coreos-cloudinit is happy with this implementation?

sdemos commented:
I haven't tested it with this implementation, but cloudinit calls the binary [here](https://github.com/coreos/coreos-cloudinit/blob/3c68e2e67d2ea077e2b79383c0ad313770c62ca2/system/ssh_key.go#L36) using a configuration of flags that I've tested and it works fine. it just runs in userspace like a normal program right?

bgilbert commented:
@sdemos It does.  We should still specifically test running it from coreos-cloudinit at boot, though, in case there's breakage caused by a difference in environment variables, etc.

sdemos committed:
cargo: add lock file to repository

adding the lock file to the repository gives us more confidence in the
reproducability of the bin build.

dm0- commented on src/lib.rs:
```diff
@@ -0,0 +1,400 @@
+
+//! update-ssh-keys library
+//!
+//! this library is meant to replace github.com/coreos/update-ssh-keys
```
This line will be self-referential when merged, which will look odd.

sdemos committed:
lib: make doc comments present tense

the comments at the top of the lib.rs file were leftover from when the
project was created. Now that the project is complete and on the verge
of being merged, we should talk about it as if it exists.

sdemos commented:
I tested provisioning a machine with coreos-cloudinit with this binary in the image instead of the shell script and it worked as expected.

sdemos committed:
lib: make ssh directory configurable

for easier testing, and also for machines with non-default ssh
directories, add a configuration option for changing the ssh directory.
The command line option exists but is hidden from the help menu.

sdemos committed:
main: improve action logging

this is both to improve user friendlyness and make the output of the
rust binary more like the output of the old shell script it is
replacing. now we print out more information about what is happening,
including the fingerprints of the keys it is affecting. we also now note
that the authorized_keys file is automatically generated in a comment.

sdemos committed:
lib: don't create empty authorized_keys file

if for whatever reason there aren't any keys in the authorized_keys.d
folder, whether a user deleted the last one or because of some
unfortunate bug, report that fact and don't change the authorized_keys
file. if a user wants to delete all their authorized public keys they
won't get any help from us.

sdemos committed:
lib: don't create file if no keys are provided

if the add function is passed an empty vec of keys, we shouldn't create
an entry for them. empty files in the authorized keys directory are
reserved for disabled keys, and should only be created in that case.

sdemos committed:
test: add python testing script

this python script is pulled directly from coreos/init, before it was
deleted in coreos/init#256. it was used for testing the shell script
version of update-ssh-keys, and will be used to test the adherance of
the rust version to the behavior of the shell script version.

sdemos committed:
test: update python test script

this python test script was pulled straight from coreos/init without any
modification. it needed a few changes before it could be used against
the rust version. here are a list of the changes and their motivations

* the fingerprints are now the SHA256 varients. this is because
  ssh-keygen, which the bash version used to print the fingerprint, now
  defaults to using SHA256 fingerprinting, so thats what the bash
  version prints out now as well.
* the configuration of the ssh directory for testing purposes is now a
  flag instead of an environment variable
* the bash version would ignore bad ssh public keys when moving the
  original authorized keys file into a newly created authorized_keys.d
  folder. the rust version does validation on all ssh public keys.
* through an implementation detail of the bash version, there were a few
  commands which could be performed at the same time. the
  test_add_and_del test tried to delete an authorized keyset and add
  another authorized keyset at the same time. the current rust
  implementation expects only one action to be taken during any
  particular invocation of the binary, so this test was removed.
* the exact output of a failure to parse a public key is different
  between to bash version and the rust version. notably, the rust
  version prints the error on stderr, otherwise it is rather cosmetic
  wording changes.

as of this commit, all of the tests in this testing script pass. in the
future, we might see about moving these tests to pure rust and using the
cargo test harness.
